RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZT_vect.cc

```diff
@@ -0,0 +1,1030 @@
+
+  while (beta < betafreeze) {
+    if(useTc_){
+      update(beta, tks,y, false, rho0);
+      while(merge(y, beta)){update(beta, tks, y, false, rho0);}
+      split(beta, tks, y);
+      beta=beta/coolingFactor_;
+    }else{
+      beta=beta/coolingFactor_;
+      splitAll(y);
+    }
+    
+    // make sure we are not too far from equilibrium before cooling further
+    niter = 0;
+    while ((update(beta, tks, y, false, rho0) > 1.e-6) && 
+	   (niter++ < maxIterations_)) {}
+
+    if(verbose_){ dump( beta, y, tks, 0); }
+  }
+  
+
+  if(useTc_){
+    //last round of splitting, make sure no critical clusters are left
+#ifdef VI_DEBUG
+    if(verbose_){ std::cout << "last spliting at " << 1./beta << std::endl; }
+#endif
+    update(beta, tks,y, false, rho0);// make sure Tc is up-to-date
+    while(merge(y,beta)){update(beta, tks,y, false, rho0);}
+    unsigned int ntry=0;
+    double threshold = 1.0;
+    while( split(beta, tks, y, threshold) && (ntry++<10) ){
+      niter=0; 
+      while((update(beta, tks,y, false, rho0)>1.e-6)  && (niter++ < maxIterations_)){}
+      while(merge(y,beta)){update(beta, tks,y, false, rho0);}
+#ifdef VI_DEBUG
+      if(verbose_){ 
+	std::cout << "after final splitting,  try " <<  ntry << std::endl; 
+	dump(beta, y, tks, 2); 
+      }
+#endif
+      // relax splitting a bit to reduce multiple split-merge cycles of the same cluster
+      threshold *= 1.1; 
+    }
+  }else{
+    // merge collapsed clusters 
+    while(merge(y,beta)){update(beta, tks,y, false, rho0);}  
+  }
+  
+#ifdef VI_DEBUG
+  if (verbose_) {
+    update(beta, tks,y, false, rho0);
+    std::cout  << "dump after 1st merging " << endl;
+    dump(beta, y, tks, 2);
+  }
+#endif
+  
+  
+  // switch on outlier rejection at T=minT
+  if(dzCutOff_ > 0){
+    rho0 = 1./nt;
+    for(unsigned int a=0; a<10; a++){ update(beta, tks, y, true, a*rho0/10);} // adiabatic turn-on
+  }
+
+  niter=0;
+  while ((update(beta, tks, y, true, rho0) > 1.e-8) && (niter++ < maxIterations_)) {};
+#ifdef VI_DEBUG
+  if (verbose_) {
+    std::cout  << "dump after noise-suppression, rho0=" << rho0  << "  niter = " << niter << endl;
+    dump(beta, y, tks, 2);
+  }
+#endif
+
+  // merge again  (some cluster split by outliers collapse here)
+  while (merge(y, beta)) {update(beta, tks, y, true, rho0); }
+#ifdef VI_DEBUG
+  if (verbose_) {
+    std::cout  << "dump after merging " << endl;
+    dump(beta, y, tks, 2);
+  }
+#endif
+
+  // go down to the purging temperature (if it is lower than tmin)
+  while( beta < betapurge_ ){
+    beta = min( beta/coolingFactor_, betapurge_);
+    niter = 0;
+    while ((update(beta, tks, y, false, rho0) > 1.e-8) && (niter++ < maxIterations_)) {}
+  }
+
+
+  // eliminate insigificant vertices, this is more restrictive at higher T
+  while (purge(y, tks, rho0, beta)) {
+    niter = 0;
+    while (( update(beta, tks, y, true, rho0) >  2.5e-7 * y.getSize() ) && (niter++ < maxIterations_)) {
+    }
```

Comment:
the same part in the Z_vect has a much simpler  ```(update(beta, tks, y, true, rho0) > 1.e-6```
is this clearly a feature of ZT clustering?
If not, perhaps it could be made configurable as well.

Reply:
This controls the time needed when the number of vertices gets very large (which depends on the minimum temperature you cool to). It is likely we could go back to the 1e-6 with a study on the annealing properties. Right now this keeps the same tolerance as Z_vect for O(100) vertices)

Reply:
Essentially this lets me speed through incorrect solutions with overly larger vertex multiplicity.